chain/messagepool/selection.go

```diff
@@ -46,24 +51,155 @@ func (mp *MessagePool) SelectMessages(ctx context.Context, ts *types.TipSet, tq
 	// if the ticket quality is high enough that the first block has higher probability
 	// than any other block, then we don't bother with optimal selection because the
 	// first block will always have higher effective performance
+	var sm *selectedMessages
+	var err error
 	if tq > 0.84 {
-		msgs, err = mp.selectMessagesGreedy(ctx, mp.curTs, ts)
+		sm, err = mp.selectMessagesGreedy(ctx, mp.curTs, ts)
 	} else {
-		msgs, err = mp.selectMessagesOptimal(ctx, mp.curTs, ts, tq)
+		sm, err = mp.selectMessagesOptimal(ctx, mp.curTs, ts, tq)
 	}
 
 	if err != nil {
 		return nil, err
 	}
 
-	if len(msgs) > build.BlockMessageLimit {
-		msgs = msgs[:build.BlockMessageLimit]
+	if sm == nil {
+		return nil, nil
+	}
+
+	// one last sanity check
+	if len(sm.msgs) > build.BlockMessageLimit {
+		sm.msgs = sm.msgs[:build.BlockMessageLimit]
```

Comment:
Log warning or error because this means that the code is wrong.

Reply:
Yessir!